[Ryan Tian] iP

ryantianj opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
* Text-based
* *Easy* to learn
* **Perfect** for people who can type fast!
* ~~Fast~~ ***Super* Fast**

To install duke:
1. Download it from [here](https://github.com/ryantianj/ip/releases/download/v0.1/Duke.jar)
2. Double-click on it
3. Add tasks
4. Let Duke manage your Tasks for you! :smile:

Features:
- [x] Manage Tasks
- [x] Search Tasks
- [x] Delete Tasks
- [x] Add Deadlines, or Event!
- [ ] Reminders (coming soon!)
- [x] More Features to Come!    

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Duke {
    public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
}

```

ryantianj committed:
level 1

ryantianj committed:
commiting git-ignore

ryantianj committed:
level 2: add, List

ryantianj committed:
Level 3: Mark as done

ryantianj committed:
Level 4: ToDos, Deadlines and Events

ryantianj committed:
A-TextUiTesting: Added simple inputs, expected outputs

ryantianj committed:
Level 5: handle errors

ryantianj committed:
Added Java Docs and Comments for v1.6

ryantianj committed:
Level 6: delete

ryantianj committed:
Added Enums

ryantianj committed:
Updated Test cases

ryantianj committed:
Updated test cases, Bug fixes

ryantianj committed:
Updated Test cases

ryantianj committed:
Update all javadoc comments

ryantianj committed:
Add Save feature

ryantianj committed:
Add date parser

ryantianj committed:
reverting to pre merge

ryantianj committed:
revert to pre merge

ryantianj committed:
Revert "Update all javadoc comments"

This reverts commit 27c171f5ae094f2b5b7d133b47d8b362f79a2cc5.

ryantianj committed:
Revert to pre branch merge

ryantianj committed:
Revert to pre merge

ryantianj committed:
Add Save feature for list history

ryantianj committed:
Merge level 7 branch

ryantianj committed:
Revert branch 7

ryantianj committed:
Revert to pre branch merge

ryantianj committed:
Revert to pre branch merge

ryantianj committed:
Merge branch 'branch-Level-7':

ryantianj committed:
Merge branch 'branch-Level-8'

ryantianj committed:
Revert to pre branch creation

ryantianj committed:
Add Save feature

ryantianj committed:
Add date parser

ryantianj committed:
Merge branch 'branch-Level-7'

ryantianj committed:
Merge branch-Level-8

ryantianj committed:
Update date parser

ryantianj committed:
Restructure program files

ryantianj committed:
Reorganise Duke into a Package

ryantianj committed:
Add JUnit tests

ryantianj committed:
Add JAR file to release

ryantianj committed:
Update Javadoc

ryantianj committed:
Update Coding Standards

ryantianj committed:
Add Find Feature

ryantianj committed:
Merge branch 'branch-A-JavaDoc'

ryantianj committed:
Merge branch 'branch-A-CodingStandard'

ryantianj committed:
Merge branch 'branch-Level-9'

ryantianj committed:
Added JavaDoc for Search

ryantianj committed:
Update access modifiers

ryantianj committed:
Add Gradle Support

ryantianj committed:
Merge branch 'branch-A-Gradle'

ryantianj committed:
Added checkstyle

ryantianj committed:
Add GUI for Duke

ryantianj committed:
Merge branch 'branch-Level-10'

ryantianj committed:
Add JavaDoc

ryantianj committed:
Update git ignore to remove lib files

ryantianj committed:
Remove lib folders from repo

ong-hongkai commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    /**
+     * Prints past commands.
+     *
+     * @return returns a printout of all the past user commands.
+     */
+    public static String userCommands() {
```
Would it be better if you chose a more intuitive name for this method?

ryantianj replied:
Agreed! Will change to listCommand

ong-hongkai commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    private void executeCommand(Parser.Command commandType) {
+        try {
+            if (commandType == Parser.Command.LIST) {
+                this.commandResult = TaskList.userCommands();
+            } else if (commandType == Parser.Command.DONE) {
+                int taskNumberDone = parsedCommand.getTaskNumber();
+                this.commandResult = commands.markDone(taskNumberDone);
+            } else if (commandType == Parser.Command.TODO) {
+                Task taskTodo = new ToDo(currentCommand.substring(5).trim(), Task.Type.T);
+                Ui.commands.add(taskTodo);
+                this.commandResult = printTask(taskTodo);
+            } else if (commandType == Parser.Command.DEADLINE) {
+                Task taskDeadline = new DeadLine(currentCommand.substring(9).trim(),
+                        Task.Type.D, returnDeadline(currentCommand).trim());
+                Ui.commands.add(taskDeadline);
+                this.commandResult = printTask(taskDeadline);
+            } else if (commandType == Parser.Command.EVENT) {
+                Task taskEvent = new Event(currentCommand.substring(6).trim(),
+                        Task.Type.E, returnTimeline(currentCommand).trim());
+                Ui.commands.add(taskEvent);
+                this.commandResult = printTask(taskEvent);
+            } else if (commandType == Parser.Command.DELETE) {
+                int taskNumberDelete = parsedCommand.getTaskNumber();
+                this.commandResult = commands.deleteTask(taskNumberDelete);
+            } else if (commandType == Parser.Command.FIND) {
+                String query = parsedCommand.getQuery();
+                this.commandResult = commands.search(query);
+            } else {
+                // Unknown Command, throw error
+                throw new DukeException("Error, Invalid Command" + "\n ");
```
I like the way you used enums here

ong-hongkai commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @return returns the String representation of a DukeException.
+     */
+    @Override
+    public String toString() {
+        if (type == 0) {
+            return "The " + message + " Command and Description must be separated by whitespace!";
+        } else if (type == 1) {
+            return "The description of " + message + " cannot be empty!";
+        } else if (type == 2) {
+            return message + " Invalid Number, only numbers greater than 0 are accepted";
+        }
+        return message;
+    }
+}
```
I like the way you implemented this.

ong-hongkai reviewed (commented):
Overall, a great project. I wasn't able to pick apart much from the project at all. The only thing I was able to spot was a potentially better choice of method name

jiarong15 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,69 @@
+     * @param type Task type: Event.
+     * @param by Event timeline.
+     */
+    public Event(String description, Type type, String by) {
```
very small issue, but maybe consider changing input String by to at? Less confusing and more consistent

ryantianj replied:
Agreed! I have changed it to at.

jiarong15 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,207 @@
+            return false;
+        }
+    }
+}
```
Perhaps you could abstract a common private method to handle the different type of tasks? They seem like repeat code. The abstracted method can take in input of length of the command(Integer), name("deadline", event")

ryantianj replied:
Agreed, will include that in the future when I review my code. Thanks

jiarong15 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,148 @@
+            String description = task.getDescription();
+            String at = task.getTime();
+            toSave += description + " | " + at;
+        } else {
```
Is there some issue here with the else statement??

ryantianj replied:
Forgot to remove it. Thanks for spotting it

jiarong15 reviewed (commented):
Generally well documented

ryantianj committed:
Update GUI to chat box

ryantianj committed:
Fix Code following PR review

ryantianj committed:
Update Gradle file for JAR creation

ryantianj committed:
Add assertions at important locations

ryantianj committed:
Fix Code according to code quality standards

Scanned through all files, fixing any coding violations
and formatting code to match coding standards.

Also removed unneccesary variables and refactored some variable
names to match the coding standards. The changes were also in line
with the pull request comments made on github.

ryantianj committed:
Add assertions throughout the code

Scanned through all the files and added assertions where
necessary to document important assumptions that should
be held at various points in the code.

Utilising assertions will aid in verifying the correctness
of the program, since the code is checked at critical points.

More pertinent locations requiring assertions:
1. Storage
2. Parser
3. TaskList
4. Duke

ryantianj committed:
Merge pull request #1 from ryantianj/branch-A-Assertions

Branch a assertions

ryantianj committed:
Merge branch 'master' int branch-A-CodeQuality

ryantianj committed:
Merge pull request #2 from ryantianj/branch-A-CodeQuality

Fix Code according to code quality standards

ryantianj committed:
Add better search functionality in Duke

As the user continues to add tasks to the list, the list can become
more difficult to read and the ability to search will enhance user
experience.

Find command:
1. Make matching case-insensitive
2. Can search through substrings
3. Can search based on task time (deadlines and events)
4. Can search tasks that have not been done yet (-d tag)
5. Can search tasks by their types using tags (-t tag, then t|d|e)

These features will make Duke more user-friendly as the user might
not always remember the casing of their tasks, or exact words that they
might have used.

Let's:
* update the search algorithms to include these features

ryantianj committed:
Merge branch 'branch-C-BetterSearch'

ryantianj committed:
Update GUI and rename bot to Neo

ryantianj committed:
Finalize Features, fix all stylings

ryantianj committed:
Update README.md

ryantianj committed:
Add product screenshot

ryantianj committed:
Remove project template

ryantianj committed:
Update Gradle for final release

ryantianj committed:
Update README.md
